File: src/main/java/com/vah/let/sort/InsertSort.java
package com.vah.let.sort;

/**
 * @Description 插入排序
 * @Author vahuang
 * @Date 2025/6/24 17:36
 **/
public class InsertSort {
    public static void main(String[] args) {
        int[] arr = {5, 2, 3, 1, 4, 6, 7, 9, 8};
//        left = 0; right = 4
//        int[] arr = {4, 2, 3, 1, 4, 6, 7, 9, 8};
//        left = 4 right = 4
//        int[] arr = {4, 2, 3, 1, 5, 6, 7, 9, 8};

        insertSort(arr);
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
    }

   public static void insertSort(int [] arr) {
        for (int i= 1; i < arr.length; i++) {
            int insertValue = arr[i];
            int insertIndex = i;
            while (insertIndex - 1 >= 0 && insertValue < arr[insertIndex - 1]) {
                arr[insertIndex] = arr[insertIndex - 1];
                insertIndex--;
            }
            arr[insertIndex] = insertValue;
        }
   }
}

File: src/main/kotlin/com/vah/let/algorithm/sort/InsertSort.kt
package com.vah.let.algorithm.sort;

/**
 *@Description 插入排序
 *
 *@Author HuangJiang
 **/
class InsertSort {
}

fun main() {
    sort(intArrayOf(8, 9, 1, 7, 2, 3, 5, 4, 6, 0))

}

private fun sort(arr: IntArray) {
    var count = 0
    for (i in 1 until arr.size) {
        // 从第二个元素开始向前插入
        val insertV = arr[i]
        var insertIndex = i
        while (insertIndex - 1 >= 0 && insertV < arr[insertIndex - 1]) {
            // 将前面的元素依次往后面移动
            arr[insertIndex] = arr[insertIndex - 1]
            insertIndex--
            count++
        }
        arr[insertIndex] = insertV
    }
    arr.forEach { print(it) }
    println()
    println(count)
}


